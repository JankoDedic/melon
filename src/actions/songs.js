import uuid from 'uuid';

export const addSong = (song) => {
  const { title = '', artists = '' } = song;
  return {
    type: 'ADD_SONG',
    song: { id: uuid.v4(), title, artists },
  };
};

export const editSong = (id, updates) => ({
  type: 'EDIT_SONG',
  id,
  updates,
});
