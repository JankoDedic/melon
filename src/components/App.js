import React from 'react';
import { connect } from 'react-redux';

import AppRouter from '../routers/AppRouter';
import Player from './Player';

export const App = ({ isUserAuthenticated }) => (
  <div>
    <AppRouter />
    {isUserAuthenticated && <Player />}
  </div>
);

const mapStateToProps = (state) => ({
  isUserAuthenticated: !!state.auth.uid,
});

export default connect(mapStateToProps)(App);
