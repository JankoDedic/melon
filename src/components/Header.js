import React from 'react';

export default () => (
  <div className="header-container">
    <div className="header">
      <div className="header__logo">
        Melon
      </div>
      <button className="header__logout">Logout</button>
    </div>
  </div>
);
