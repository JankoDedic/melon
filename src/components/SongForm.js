import React from 'react';
import ReactPlayer from 'react-player';
import PropTypes from 'prop-types';

export class SongForm extends React.Component {
  constructor(props) {
    super(props);

    if (props.song) {
      this.state = { ...props.song, error: '' };
    } else {
      this.state = {
        title: '',
        artists: '',
        url: '',
        error: '',
      };
    }
  }
  handleTitleChange = (e) => {
    this.setState({
      title: e.target.value,
    });
  }
  handleArtistsChange = (e) => {
    this.setState({
      artists: e.target.value,
    });
  }
  handleURLChange = (e) => {
    this.setState({
      url: e.target.value,
    });
  }
  handleSubmit = (e) => {
    e.preventDefault();
    const { id, title, artists, url } = this.state;
    const isTitleValid = title.length > 0;
    const isURLValid = ReactPlayer.canPlay(url);
    if (!isTitleValid) {
      this.setState({ error: 'Invalid title.' });
    } else if (!isURLValid) {
      this.setState({ error: 'Invalid URL. Melon cannot play this media.' });
    } else {
      this.props.onSubmit({ title, artists, url });
    }
  }
  handleRemove = (e) => {
    e.preventDefault();
    this.props.onRemove();
  }
  render() {
    const { removeButton } = this.props;
    return (
      <div>
        <div className="song-form__header">
          {this.props.formTitle}
        </div>
        {this.state.error && <p className="song-form__error">{this.state.error}</p>}
        <form className="song-form__form" onSubmit={this.handleSubmit}>
          <input
            type="text"
            spellCheck="false"
            placeholder="Title"
            autoFocus
            value={this.state.title}
            onChange={this.handleTitleChange}
          />
          <input
            type="text"
            spellCheck="false"
            placeholder="Artists"
            value={this.state.artists}
            onChange={this.handleArtistsChange}
          />
          <input
            type="text"
            spellCheck="false"
            placeholder="URL (YouTube, SoundCloud...)"
            value={this.state.url}
            onChange={this.handleURLChange}
          />
          <div className="song-form__form__buttons">
            <button className="primary-button">Save Song</button>
            {removeButton && <button onClick={this.handleRemove}>Remove Song</button>}
          </div>
        </form>
      </div>
    );
  }
}

SongForm.propTypes = {
  title: PropTypes.string,
  onSubmit: PropTypes.func,
  song: PropTypes.shape({
    title: PropTypes.string,
    artists: PropTypes.string,
    url: PropTypes.string,
  }),
};

export default SongForm;
