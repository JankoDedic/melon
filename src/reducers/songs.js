const defaultSongs = [
  { title: 'This Is What You Came For', artists: 'Calvin Harris, Rihanna' },
  { title: 'Your Way', artists: 'Frank Sinatra' }
];

export default (songs = defaultSongs, action) => {
  return songs;
};

