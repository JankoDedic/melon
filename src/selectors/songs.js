export default (songs, filters) => {
  return songs.filter((song) => {
    const doesTitleMatch = song.title.toLowerCase().includes(filters.text.toLowerCase());
    const doArtistsMatch = song.artists.toLowerCase().includes(filters.text.toLowerCase());
    return doesTitleMatch || doArtistsMatch;
  });
};
