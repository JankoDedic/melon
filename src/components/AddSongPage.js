import React from 'react';
import { connect } from 'react-redux';

import SongForm from './SongForm';
import { startAddSong } from '../actions/songs';

class AddSongPage extends React.Component {
  handleSubmit = (song) => {
    this.props.startAddSong(song);
    this.props.history.push('/dashboard');
  }
  render() {
    return (
      <SongForm formTitle="Add Song" onSubmit={this.handleSubmit} />
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddSong: (song) => dispatch(startAddSong(song)),
});

export default connect(undefined, mapDispatchToProps)(AddSongPage);
