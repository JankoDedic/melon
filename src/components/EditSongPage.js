import React from 'react';
import { connect } from 'react-redux';

import SongForm from './SongForm';
import { startEditSong, removeSong } from '../actions/songs';

class EditSongPage extends React.Component {
  handleSubmit = (song) => {
    this.props.startEditSong(this.props.song.id, song);
    this.props.history.push('/dashboard');
  }
  handleRemove = () => {
    this.props.removeSong(this.props.song.id);
    this.props.history.push('/dashboard');
  }
  render() {
    return (
      <div>
        <SongForm
          formTitle="Edit Song"
          onSubmit={this.handleSubmit}
          song={this.props.song}
        />
        <button
          className="form-button"
          onClick={this.handleRemove}
        >
          Remove Song
        </button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  song: state.songs.find((song) => song.id === props.match.params.id),
});

const mapDispatchToProps = (dispatch, ownProps) => ({
  startEditSong: (id, updates) => dispatch(startEditSong(id, updates)),
  removeSong: (id) => dispatch(removeSong(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditSongPage);
