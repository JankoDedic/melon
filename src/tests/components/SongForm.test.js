import React from 'react';
import { shallow } from 'enzyme';

import { SongForm } from '../../components/SongForm';
import songs from '../fixtures/songs';

test('SongForm snapshot with default song info', () => {
  const wrapper = shallow(<SongForm />);
  expect(wrapper).toMatchSnapshot();
});

test('SongForm snapshot', () => {
  const wrapper = shallow(<SongForm song={songs[1]} />);
  expect(wrapper).toMatchSnapshot();
});

test('SongForm default state', () => {
  const wrapper = shallow(<SongForm />);
  expect(wrapper.state()).toEqual({
    title: '',
    artists: '',
    url: '',
    error: ''
  });
});

describe('SongForm inputs', () => {
  let e, wrapper;

  beforeAll(() => {
    e = { target: { value: 'sample text here' }};
    wrapper = shallow(<SongForm />);
  });

  test('song title input handles text change', () => {
    wrapper.find('input').at(0).simulate('change', e);
    expect(wrapper.state('title')).toBe(e.target.value);
  });

  test('song artists input handles text change', () => {
    wrapper.find('input').at(1).simulate('change', e);
    expect(wrapper.state('artists')).toBe(e.target.value);
  });

  test('song url input handles text change', () => {
    wrapper.find('input').at(2).simulate('change', e);
    expect(wrapper.state('url')).toBe(e.target.value);
  });
});

describe('given valid song information', () => {
  const song = songs[2];
  const onSubmit = jest.fn();
  const wrapper = shallow(<SongForm onSubmit={onSubmit} song={song} />);

  describe('given a form submit event', () => {
    const ev = { preventDefault: () => {} };
    wrapper.find('form').simulate('submit', ev);

    it('calls onSubmit with song info', () => {
      const { id, ...songInfo } = song;
      expect(onSubmit).toHaveBeenLastCalledWith({ ...songInfo });
    });
  });
});

describe('given song info with invalid title', () => {
  const song = {
    ...songs[1],
    title: ''
  };
  const onSubmit = jest.fn();
  const wrapper = shallow(<SongForm onSubmit={onSubmit} song={song} />);

  describe('given a form submit event', () => {
    const ev = { preventDefault: () => {} };
    wrapper.find('form').simulate('submit', ev);

    it('does not call onSubmit', () => {
      const { id, ...songInfo } = song;
      expect(onSubmit).not.toHaveBeenCalled();
    });

    it('sets the error state', () => {
      expect(wrapper.state('error')).not.toBeFalsy();
      expect(wrapper).toMatchSnapshot();
    });
  });
});

describe('given song info with invalid URL', () => {
  const song = {
    ...songs[0],
    url: ''
  };
  const onSubmit = jest.fn();
  const wrapper = shallow(<SongForm onSubmit={onSubmit} song={song} />);

  describe('given a form submit event', () => {
    const ev = { preventDefault: () => {} };
    wrapper.find('form').simulate('submit', ev);

    it('does not call onSubmit', () => {
      const { id, ...songInfo } = song;
      expect(onSubmit).not.toHaveBeenCalled();
    });

    it('sets the error state', () => {
      expect(wrapper.state('error')).not.toBeFalsy();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
