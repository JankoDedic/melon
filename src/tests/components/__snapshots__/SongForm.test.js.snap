// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongForm snapshot 1`] = `
<div>
  <div
    className="song-form__header"
  />
  <form
    className="song-form__form"
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Title"
      spellCheck="false"
      type="text"
      value="Knights of Cydonia"
    />
    <input
      onChange={[Function]}
      placeholder="Artists"
      spellCheck="false"
      type="text"
      value="Muse"
    />
    <input
      onChange={[Function]}
      placeholder="URL (YouTube, SoundCloud...)"
      spellCheck="false"
      type="text"
      value="https://www.youtube.com/watch?v=G_sBOsh-vyI"
    />
    <div
      className="song-form__form__buttons"
    >
      <button
        className="primary-button"
      >
        Save Song
      </button>
    </div>
  </form>
</div>
`;

exports[`SongForm snapshot with default song info 1`] = `
<div>
  <div
    className="song-form__header"
  />
  <form
    className="song-form__form"
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Title"
      spellCheck="false"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Artists"
      spellCheck="false"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="URL (YouTube, SoundCloud...)"
      spellCheck="false"
      type="text"
      value=""
    />
    <div
      className="song-form__form__buttons"
    >
      <button
        className="primary-button"
      >
        Save Song
      </button>
    </div>
  </form>
</div>
`;

exports[`given song info with invalid URL given a form submit event sets the error state 1`] = `
<div>
  <div
    className="song-form__header"
  />
  <p
    className="song-form__error"
  >
    Invalid URL. Melon cannot play this media.
  </p>
  <form
    className="song-form__form"
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Title"
      spellCheck="false"
      type="text"
      value="Feeling Good"
    />
    <input
      onChange={[Function]}
      placeholder="Artists"
      spellCheck="false"
      type="text"
      value="Muse"
    />
    <input
      onChange={[Function]}
      placeholder="URL (YouTube, SoundCloud...)"
      spellCheck="false"
      type="text"
      value=""
    />
    <div
      className="song-form__form__buttons"
    >
      <button
        className="primary-button"
      >
        Save Song
      </button>
    </div>
  </form>
</div>
`;

exports[`given song info with invalid title given a form submit event sets the error state 1`] = `
<div>
  <div
    className="song-form__header"
  />
  <p
    className="song-form__error"
  >
    Invalid title.
  </p>
  <form
    className="song-form__form"
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Title"
      spellCheck="false"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Artists"
      spellCheck="false"
      type="text"
      value="Muse"
    />
    <input
      onChange={[Function]}
      placeholder="URL (YouTube, SoundCloud...)"
      spellCheck="false"
      type="text"
      value="https://www.youtube.com/watch?v=G_sBOsh-vyI"
    />
    <div
      className="song-form__form__buttons"
    >
      <button
        className="primary-button"
      >
        Save Song
      </button>
    </div>
  </form>
</div>
`;
